File: animray/Examples/mandelbrot/mandelbrot.cpp
/*
    Copyright 2010, Kirit Saelensminde.
    http://www.kirit.com/AnimRay

    This file is part of AnimRay.

    AnimRay is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AnimRay is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with AnimRay.  If not, see <http://www.gnu.org/licenses/>.
*/


#include <fost/unicode>
#include <fost/main>
#include <fost/coerce/ints.hpp>
#include <animray/targa.hpp>
#include <animray/mandelbrot.hpp>


FSL_MAIN(
    L"mandelbrot",
    L"Mandelbrot, Copyright 2010 Kirit Saelensminde"
)( fostlib::ostream &out, fostlib::arguments &args ) {
    boost::filesystem::wpath output_filename =
        fostlib::coerce< boost::filesystem::wpath >(args[1].value("out.tga"))
    ;
    int width = fostlib::coerce< int >( args[2].value("100") );
    int height = fostlib::coerce< int >( args[3].value("100") );
    out << "Creating image " << output_filename
        <<", size " << width << " x " << height << std::endl
    ;

    typedef animray::film< uint8_t > film_type;
    film_type output(width, height);

    typedef double precision;
    precision centre_x = fostlib::coerce< precision >(
        args.commandSwitch("cx").value("0")
    );
    precision centre_y = fostlib::coerce< precision >(
        args.commandSwitch("cy").value("0")
    );
    precision radius = fostlib::coerce< precision >(
        args.commandSwitch("r").value("2")
    );
    std::size_t bits = fostlib::coerce< int >(
        args.commandSwitch("bits").value("8")
    );

    out << "Centre image at " << centre_x << ", " << centre_y <<
        " with radius of " << radius <<
        " to " << bits << " bits" <<
    std::endl;

    output.transform(
        animray::mandelbrot::iterations< film_type, precision >(
            width, height, centre_x, centre_y, radius, bits
        )
    );

    animray::targa(output_filename, output);

    return 0;
}

File: animray/Examples/mandelbrot/mandelbrot.py
#!/usr/local/bin/python
# -*- coding: utf-8 -*-
"""
    Copyright 2010, Kirit Saelensminde.
    http://www.kirit.com/AnimRay

    This file is part of AnimRay.

    AnimRay is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AnimRay is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with AnimRay.  If not, see <http://www.gnu.org/licenses/>.
"""

import os, sys
import pygame
from pygame.locals import *

if not pygame.font: print 'Warning, fonts disabled'
if not pygame.mixer: print 'Warning, sound disabled'


def generate(screen, x, y, radius):
    """
    """
    from animray import film_gray8, mandelbrot

    size = screen.get_size()
    film = film_gray8(size[0], size[1], 0)
    mandelbrot(film, x, y, radius, 8)

    return film


def draw(screen, cx, cy, r):
    """
    """
    film = generate(screen, cx, cy, r)
    background = pygame.Surface(screen.get_size())
    for x in range(0, film.width):
        for y in range(0, film.height):
            g = film(x, y)
            background.set_at((x, y), pygame.Color(g, g, g))
    screen.blit(background, (0, 0))
    pygame.display.flip()


def main(*args, **kwargs):
    """
    """
    pygame.init()
    screen = pygame.display.set_mode((300, 300), RESIZABLE)
    pygame.display.set_caption('AnimRay Mandelbrot')

    center_x, center_y = 0, 0
    radius = 2
    draw(screen, center_x, center_y, radius)

    to_exit = False
    while not to_exit:
        event = pygame.event.wait()
        if event.type == QUIT or \
                (event.type == KEYDOWN and event.key==K_ESCAPE):
            to_exit = True
        elif event.type == VIDEORESIZE:
            screen = pygame.display.get_surface()
            screen = pygame.display.set_mode((event.w, event.h), RESIZABLE)
            draw(screen, center_x, center_y, radius)
        else:
            print event
